Add Safe Mode

derrod opened the pull request:
### Description

This PR adds several things:
- `obs_add_safe_module(name)` to allow specifying a list of safe modules, if the list is not empty, only the specified modules are allowed to load
- Automatic detection of unclean shutdowns (disabled in `Debug` builds)
- Prompt to ask user if they want to run OBS in safe or normal mode when an unclean shutdown has been detected
- `--safe-mode` flag to run OBS in safe mode
- `--only-bundled-plugins` flag to run OBS without third-party plugins
- `--disable-shutdown-check` flag to disable unclean shutdown detection

"Safe Mode" currently disables third-party plugins as well as `frontend-tools` and `obs-websocket` to prevent scripts from running or external applications from making changes to OBS's state.

**Screenshot:**
![2023-03-14_06-32-37_iCsf5G](https://user-images.githubusercontent.com/3123295/224905557-02bf3b7d-e2b9-46df-a809-2fdb0bef2598.png)

The automatic crash/improper shutdown detection works via a sentinel file in the configuration directory that gets created on startup and deleted on shutdown.

### Motivation and Context

Make it possible to run OBS without third-party plugins for debugging or troubleshooting purposes without having to uninstall them.

Additionally, if OBS crashes or is otherwise shut down improperly allow the user to disable third-party plugins on next startup.

### How Has This Been Tested?

- Removed `win-capture` from the list, launched OBS, no window capture sources.
- Killed OBS via VS debugger and verified that the prompt appears on next launch

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

tytan652 commented on UI/window-basic-main.cpp:
```diff
@@ -231,6 +231,38 @@ static void AddExtraModulePaths()
 #endif
 }
 
+static const char *safe_modules[] = {
+	// Cross-platform
+	"aja", "aja-output-ui", "decklink", "decklink-captions",
+	"decklink-output-ui", "frontend-tools", "image-source", "obs-browser",
+	"obs-browser-page", "obs-ffmpeg", "obs-filters", "obs-libfdk",
+	"obs-outputs", "obs-transitions", "obs-vst", "obs-websocket",
+	"obs-x264", "rtmp-services", "text-freetype2", "vlc-video",
+#ifdef _WIN32
+	// Windows
+	"coreaudio-encoder", "obs-qsv11", "obs-text", "win-capture",
+	"win-dshow", "win-wasapi",
+#elif defined(__APPLE__)
+	// macOS
+	"coreaudio-encoder", "mac-avcapture", "mac-capture", "mac-syphon",
+	"mac-videotoolbox", "mac-virtualcam",
+#else
+	// Linux (and BSDs...)
+	"linux-alsa", "linux-capture", "linux-jack", "linux-pipewire",
+	"linux-pulseaudio", "linux-v4l2", "oss-audio", "sndio",
+#endif
+	NULL};
```
This should be generated through CMake in a generated header.
Maybe a macro with all the string and add it to `safe_modules`, and the macro is CMake generated.

derrod replied:
Updated to create the list from the CMake `OBS_MODULES_LIST` variable.

derrod commented:
Updated the PR now to:
- Use CMake to generate the list of safe modules (based on modules enabled at build time)
- Simplify check for unclean shutdown to simply check for existence of the sentinel file
- Change warning to a prompt allowing the user to choose between running in normal or safe mode
- Added `--unsafe` parameter to disable the automatic check (mostly useful for development, i.e. myself)
- Disabled the automatic check in `Debug` builds
- Updated wording and logging messages

notr1ch commented on UI/data/locale/en-US.ini:
```diff
@@ -124,6 +124,12 @@ AlreadyRunning.Title="OBS is already running"
 
+# warning if auto safe-mode has engaged
+AutoSafeMode.Title="Safe Mode"
+AutoSafeMode.Text="OBS did not shut down properly during your last session.\n\nWould you like to start in Safe Mode and disable all third party plugins?"
```
We could probably nitpick this text all day, but this suggests running in safe mode and disabling 3rd party plugins are two separate actions. Perhaps something like "Would you like to start in Safe Mode (disables third party plugins)?"

derrod replied:
This wording was suggested by Warchamp but I agree that could be clearer that those are the same thing.

notr1ch reviewed (commented):
Not sure if I missed it in the review, but this doesn't look like it checks if the user is using any 3rd party modules before presenting the safe mode option. Ideally we would skip the dialog in such a situation as regular and safe mode would behave identically, only causing user confusion.

derrod commented:
> Not sure if I missed it in the review, but this doesn't look like it checks if the user is using any 3rd party modules before presenting the safe mode option. Ideally we would skip the dialog in such a situation as regular and safe mode would behave identically, only causing user confusion.

I'm not sure what would be a good way to do that cross-platform, but it's a valid complaint. However, since third-party plugins are (at least anecdotally) the main reason users see crashes this may still be right more often than not.

I was originally considering also disabling more things (e.g. scripting) in this mode to further eliminate potential issues. Perhaps we could also disable some other non-essential features that we think could cause issues?

derrod commented:
Based on the off-thread conversation on Discord I gleamed that the main feedback is:

- Safe mode should do more than just disable third-party plugins
- We may want to provide a separate option for "Safe Mode" and "Run without third-party plugins"

My thoughts on that would be to add a second list of "volatile" first-party plugins that will be skipped in safe mode. For instance this may include `frontend-tools` (includes scripting) and `obs-browser` (on macOS), potentially also things like `obs-webscket` that are non-essential.

In this implementation safe mode would provide a barebones OBS instance that can be used to export profiles/collections, try deleting sources/filters that could cause issues, etc. This could allow non-technical users to more easily recover their OBS setup or at least back things up.

derrod commented:
Rework based on feedback. Now there are two separate modes:
- Safe Mode
- No third-party plugins

The former can now also block first-party plugins that could be problematic (e.g. scripting, browser) and can be queried from anywhere in the UI via `App()->IsSafeMode()`. So that should make it easier to add other limitations if we so choose.

Unfortunately this made the dialog a bit unweildy:
![2023-03-13_02-20-16_SASJph](https://user-images.githubusercontent.com/3123295/224588393-a40d90da-a52b-4b45-badb-c2d7aa3da842.png)

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -517,6 +517,13 @@ source_group(
 unset(_UI)
 
+get_property(OBS_MODULE_LIST GLOBAL PROPERTY OBS_MODULE_LIST)
+foreach(module ${OBS_MODULE_LIST})
```
Use `list(JOIN ..)` instead (https://cmake.org/cmake/help/latest/command/list.html?highlight=list#join) - I guess `\n\t` as glue should work.

derrod replied:
Would that quote each element though?

Alternatively I suppose it would be possible to make it one comma separated string and split it at runtime rather than relying on CMake to produce a full list.

PatTheMav replied:
In that case you'd output `"<glued string>"` and make `"\n\t"` the glue.

If we pass one long string, we wouldn't need a separate header and you could pass it as a compiler definition, though I'm afraid we could run afoul of Windows command line length limits then?

derrod replied:
The limit is 8000-something and a list of all modules is "only" around 350 so we probably still have a decent bit of a breathing room there. I'll try it at least.

derrod replied:
I've now reworked it to use a define. Seems to work fine on my Windows system.

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -517,6 +517,13 @@ source_group(
+foreach(module ${OBS_MODULE_LIST})
+  set(SAFE_MODULES "\t\"${module}\",\n${SAFE_MODULES}")
+endforeach()
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/modules.hpp.in
```
In general I want to remove almost all usage of generated headers in the project as they trigger rebuilds too often (the headers are regenerated by configuring the project, thus invalidating every binary that depends on them) - in this case almost the entire `obs` target will be rebuilt.

If there is a cleaner way to pass this information into the code without relying on generated files, I'd rather use that.

derrod replied:
The only other way I could really think of was to use a define, but having a long, quoted, and comma seprated string in a single define felt somewhat dirty/risky.

PatTheMav replied:
Yeah I’d have CMake generate straight C code (so no defines), but we can only change that in bulk at a later time, similar to the version library.

PatTheMav commented on UI/modules.hpp.in:
```diff
@@ -0,0 +1,13 @@
+
+static const std::unordered_set<std::string> unsafe_modules = {
+#ifdef __APPLE__
+	"obs-browser", // Known to occasionally cause issues on macOS
```
I don't think we should remove obs-browser - if it is an "unsafe" module, we should fix that instead. Unfortunately that plugin has become a core functionality of OBS, as we require CEF for too many other things already.

If it is indeed unstable, the 'What's New' functionality should not rely on it and we should use `WKWebKit` on macOS instead. It's one or the other.

derrod replied:
This was added based on @gxalpha's recommendation, he might be able to elaborate more under which circumstances it causes issues.

Generally I think that we could look into moving at least non-source browser elements to something like https://github.com/webview/webview to use the native webviews on various systems.

gxalpha replied:
obs-browser was one example of a module I gave that has caused problems in the past (older versions of OBS on macOS 11 iirc, where people were recommended to just delete their scene collection to fix the problem) when advocating for a much more locked down variant of a safe mode, which would function mostly as a mode to get back out of situations where the current scene collections somehow got completely screwed. Such a mode would have greatly reduced core functionality, with obs-browser being a prime candidate of a module that should not be loaded.

All that said, the kind of Safe Mode this PR proposes is a mode without third-party plugins and scripts, and I'm definitely not in favor of disabling obs-browser there. My communication may have been unclear in that regard.
I do think calling it "Safe Mode" may not be the greatest name for the mode that this PR proposes (operating systems generally get very locked down in their Safe Modes), but that's irrelevant to the current discussion.

derrod replied:
In that case I'll remove it and update the wording to specifically call out that only third-party plugins and scripting will be disabled.

As mentiond, you're not really supposed to use OBS in this mode, it's so you can export your data or perhaps try to delete a source/filter belonging to a plugin that is causing an issue. And while yes, you can do those things by manually copying or editing the JSON data, your average user is not going to understand that.

PatTheMav commented on UI/obs-app.cpp:
```diff
@@ -2403,6 +2415,10 @@ static int run_program(fstream &logFile, int argc, char *argv[])
+			/* Clear unclean_shutdown flag as multiple instances
+			 * running from the same config will lead to a
+			 * false-positive detection.*/
+			unclean_shutdown = false;
```
Are there still valid use cases for multiple instances? We should either fully support it, or remove support for it, but not this weird in-between state, where users can choose to put the application in a state where we cannot ensure full functionality.

PatTheMav commented on UI/obs-app.cpp:
```diff
@@ -3176,6 +3217,47 @@ static void upgrade_settings(void)
+
+static void check_safe_mode_sentinel(void)
+{
+#ifndef NDEBUG
```
I don't think we set `NDEBUG` on any platform but Windows - it is common to explicitly set `DEBUG` when it is enabled, but not set anything otherwise.

PatTheMav replied:
I wonder if it would be more efficient to guard the _function call_ behind that instead of creating an immediately returning function, though I would also suspect that modern compilers optimise that away anyway?

PatTheMav commented on UI/obs-app.cpp:
```diff
@@ -3176,6 +3217,47 @@ static void upgrade_settings(void)
+		return;
+
+	if (!os_file_exists(path)) {
+		os_quick_write_utf8_file(path, sentinelContents,
```
Is `os_quick_write_utf8_file` guaranteed to flush the buffer to disk immediately? Because otherwise this might not do anything.

derrod replied:
It does flush the file: https://github.com/obsproject/obs-studio/blob/b31344d82ec60608e03b9648f5ff96fa486d5c9a/libobs/util/platform.c#LL280C7-L280C7

PatTheMav commented on libobs/obs.h:
```diff
@@ -517,6 +517,14 @@ EXPORT const char *obs_get_module_data_path(obs_module_t *module);
+ *
+ * @param  name  Specifies the module's name (filename sans extension).
+ */
+EXPORT void obs_add_safe_module(const char *name);
```
Doesn't this make the function available to plugins as well? Meaning that _anybody_ can add safe modules to this list?

derrod replied:
Yes, that is a limitation of libobs, we don't really have a means of defining "UI only" APIs. But since third-party plugins aren't ever loaded, they cannot add themselves if that's what you were thinking.

dodgepong commented:
Perhaps out of scope, but would it be worthwhile to add a menu option somewhere in OBS to restart the app in Safe Mode?

derrod commented:
> Perhaps out of scope, but would it be worthwhile to add a menu option somewhere in OBS to restart the app in Safe Mode?

Maybe? At least for the purposes of our "pls test this without plugins" response in support channels that might not be terrible. But I'll defer to @Warchamp7 on such UI/UX questions.

dodgepong commented:
Indeed, I'm mostly thinking about the support case. Otherwise, the only way to intentionally start the app in safe mode is via a command line flag, which is not especially user-friendly.

On the other hand, users aren't typically going to want to use Safe Mode at all, so there's an argument to be made that without some amount of obvious signage when the app starts in Safe Mode, a menu item to restart in Safe Mode could turn into a footgun.

Alternatively, we could ship a second shortcut with the installer (e.g. "OBS Studio (Safe Mode)") to start the app in safe mode, though I suspect that's liable to be ignored/misunderstood/deleted/overlooked. It's also perhaps even more vulnerable to the footgun issue.

notr1ch commented:
Given Safe Mode is intended for recovery situations, would it make sense to disable features like auto start stream / recording? These may trigger crashing code paths such as during encoder initialization.

derrod commented:
> Given Safe Mode is intended for recovery situations, would it make sense to disable features like auto start stream / recording? These may trigger crashing code paths such as encoder initialization.

It does make sense, yes. I'm not sure what the best way to go about it would be yet. I think perhaps making all the UI methods such as "StartStreaming" a no-op might work, although we might want to add some messagebox explaining to the user that it is disabled while in safe mode or something.

derrod commented:
Updated now to use the existing `EnableOutputs(bool)` to disable any programmatic starts of the various outputs, and also added a little info message that pops up when clicking on the buttons:
![2023-03-14_03-24-15_7i7iSm](https://user-images.githubusercontent.com/3123295/224877342-230b7f58-f9bf-4fdc-aee6-7a774b83756c.png)

Also fixed plugin data stored in the scene collection (in `modules`) being lost when plugins are not loaded, thanks @tytan652 for making me aware of this.

Additionally, rather than using a configure file it now just passes a list as a compile definition and uses the convenient `readline()` for stringstreams to read them one by one. The "unsafe" list has now been moved back into the `window-basic-main.cpp` file.

Finally, I also added `obs-websocket` to the list of "unsafe" modules to prevent third-party software interacting with OBS through websockets from modifying the state while in Safe Mode.

derrod commented:
In probably the final update for today, also added an option to restart in safe mode to the "Help" menu:
![2023-03-14_07-51-25_Od24As](https://user-images.githubusercontent.com/3123295/224919056-34bf6e83-5a1f-4f8c-98c3-8c8dcedce06c.png)

After clicking the option this message box will appear explaining what the mode does and offering them to continue or abort:
![2023-03-14_07-23-07_HGQiS7](https://user-images.githubusercontent.com/3123295/224919073-19088b6b-e2b5-4272-a10a-6efa29020965.png)

notr1ch commented:
Small nitpick, "About" is traditionally the last item on a Help menu. This feels like it belongs more with the Check File Integrity / Check For Updates group.

derrod commented:
> Small nitpick, "About" is traditionally the last item on a Help menu. This feels like it belongs more with the Check File Integrity / Check For Updates group.

Yep, done.

TianQiBuTian commented:
Maybe disable `Restart in Safe Mode` button in safe mode. (used to restart obs in safe mode again?)
![10](https://user-images.githubusercontent.com/26628474/225135134-52302a40-3898-4384-8718-d1460c9ed959.png)
Idea: OR Change to `Restart in Normal Mode` in safe mode.

derrod commented:
Just a minor update to implement the suggested change of replacing the option in the help menu with "Restart in Normal Mode" when in safe mode and only log when the options are actually set to reduce log spam a little.

norihiro commented:
I worry that the audio track setting will become zero for the 3rd party plugins and the audio won't be streamed/recorded.
This is not the bug of this PR but the issue will be easily triggered when support volunteers ask to use this function.

The step to reproduce the issue is as below.
1. Install a 3rd party audio plugin, for example [obs-vban](https://github.com/norihiro/obs-vban/).
2. Start OBS
3. Add a source `VBAN Audio Source` (no need to configure)
4. Open Advanced Audio Properties and make sure all tracks are enabled.
5. Restart with the safe mode
6. Restart without the normal mode
7. Open Advanced Audio Properties and make sure all tracks. Expecting all tracks are enabled but actually disabled.

derrod commented:
I imagine that just needs to be saved as well somewhere. Perhaps the default it falls back to should be all tracks rather than none, but ideally we'd retain that information properly.

derrod commented:
@norihiro I think this should do it? (Haven't tested it yet)

norihiro commented:
> @norihiro I think this should do it? (Haven't tested it yet)

I tested with NDI Source and VBAN Source. Both works fine. Thank you so much.

> I'm not sure what to really call those placeholder sources created when the source type is missing, so I just went with "phantom sources".

I was calling it "undefined source" but I'm not sure it's accurate enough to describe the state of the source.

derrod commented:
Rebased on master, conflicts resolved, updated for cmake 3.0. Still works as intended.

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -108,6 +108,10 @@ foreach(graphics_library IN ITEMS opengl metal d3d11)
 endforeach()
 
+get_property(obs_module_list GLOBAL PROPERTY OBS_MODULES_ENABLED)
+list(JOIN obs_module_list "|" SAFE_MODULES)
```
I'd prefer using `configure_file` to output a source file `obs_safe_modules.c` that sets up a static array directly and also contains utility functions for that list. Gives us encapsulation, a single location to manage basic functionality, and doesn't rely on macros/preprocessor strings.

I would expect the contents of this file to be quite static, so it shouldn't hurt having it added directly as a compilation unit. If the functionality is used in UI code only, we could generate it as C++ directly even.

derrod replied:
Originally this used a generated header though based on review comments I changed it to a define (https://github.com/obsproject/obs-studio/pull/8455#discussion_r1134005599).

PatTheMav replied:
Yeah I’d have CMake generate straight C code (so no defines), but we can also change that in bulk at a later time, similar to the version library.

derrod replied:
Originally I created a header file containing a `unordered_set<string>` with the module names: https://github.com/obsproject/obs-studio/commit/7031ee85c61c33c04f3ae0b1be05dae6d728b9ac#diff-5d2ded8a0a8649fab4ca9479064e8938310c4d65522dea409764ec6b776ea25c

If that's what you mean I can simply go back to that. Otherwise yes we can do it later.

PatTheMav replied:
Damn that’s kinda what I meant, probably brainlagged in the past. But let’s go with the current implementation to get this over the finish line.

RytoEX commented on UI/data/locale/en-US.ini:
```diff
@@ -125,6 +125,15 @@ AlreadyRunning.Title="OBS is already running"
 AlreadyRunning.Text="OBS is already running! Unless you meant to do this, please shut down any existing instances of OBS before trying to run a new instance. If you have OBS set to minimize to the system tray, please check to see if it's still running there."
 AlreadyRunning.LaunchAnyway="Launch Anyway"
 
+# warning if auto safe-mode has engaged
```
Nit: Safe Mode

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -1715,6 +1720,12 @@ bool OBSApp::OBSInit()
 	blog(LOG_INFO, "Portable mode: %s", portable_mode ? "true" : "false");
 
+	if (safe_mode) {
+		blog(LOG_WARNING, "Safe mode enabled.");
```
Nit: Safe Mode

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -1715,6 +1720,12 @@ bool OBSApp::OBSInit()
+	if (safe_mode) {
+		blog(LOG_WARNING, "Safe mode enabled.");
+	} else if (disable_3p_plugins) {
+		blog(LOG_WARNING, "Third party plugins disabled.");
```
Nit: Third-party

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -2453,6 +2468,34 @@ static int run_program(fstream &logFile, int argc, char *argv[])
+			safe_mode = mb.clickedButton() == launchSafeButton;
+			if (safe_mode) {
+				blog(LOG_INFO,
+				     "[Safe Mode] User has launched in safe mode.");
```
Nit: Safe Mode

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -3194,6 +3246,32 @@ static void upgrade_settings(void)
+static void check_safe_mode_sentinel(void)
+{
+#ifndef NDEBUG
+	/* Safe mode detection is disabled in Debug builds to keep developers
```
Nit: Safe Mode

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -3347,6 +3436,9 @@ int main(int argc, char *argv[])
 				"--portable, -p: Use portable mode.\n"
 #endif
 				"--multi, -m: Don't warn when launching multiple instances.\n\n"
+				"--safe-mode: Run in safe mode (disables third-party and some built-in plugins).\n"
```
```suggestion
				"--safe-mode: Run in Safe Mode (disables third-party and some built-in plugins).\n"
```

RytoEX commented on UI/window-basic-main.cpp:
```diff
@@ -232,6 +233,30 @@ static void AddExtraModulePaths()
 #endif
 }
 
+/* First-party modules considered to be potentially unsafe to load in safe mode
```
```suggestion
/* First-party modules considered to be potentially unsafe to load in Safe Mode
```

RytoEX commented on UI/window-basic-main.cpp:
```diff
@@ -801,9 +826,18 @@ void OBSBasic::Save(const char *file)
-		api->on_save(moduleObj);
-		obs_data_set_obj(saveData, "modules", moduleObj);
+		if (safeModeModuleData) {
+			/* If we're in safe mode and have retained unloaded
```
```suggestion
			/* If we're in Safe Mode and have retained unloaded
```

RytoEX commented on UI/window-basic-main.cpp:
```diff
@@ -801,9 +826,18 @@ void OBSBasic::Save(const char *file)
-		obs_data_set_obj(saveData, "modules", moduleObj);
+		if (safeModeModuleData) {
+			/* If we're in safe mode and have retained unloaded
+			 * plugin data update the existing data object instead
```
```suggestion
			 * plugin data, update the existing data object instead
```

RytoEX commented on UI/window-basic-main.cpp:
```diff
@@ -1961,7 +2001,14 @@ void OBSBasic::OBSInit()
 	struct obs_module_failure_info mfi;
 
-	AddExtraModulePaths();
+	/* Safe mode disables third-party plugins so we don't need to add earch
```
```suggestion
	/* Safe Mode disables third-party plugins so we don't need to add earch
```

RytoEX commented on libobs/obs.h:
```diff
@@ -517,6 +517,14 @@ EXPORT const char *obs_get_module_data_path(obs_module_t *module);
 EXPORT void obs_add_module_path(const char *bin, const char *data);
 
+/**
+ * Adds a module to the list of modules allowed to load in safe mode.
```
```suggestion
 * Adds a module to the list of modules allowed to load in Safe Mode.
```

RytoEX reviewed (changes requested):
Just nits.

RytoEX reviewed (approved):
Tested on macOS. Approved from my end. Ideally, would like an approval from @notr1ch .
